### ISSUE TITLE
Confused about behaviour vs tsc behaviour

### ISSUE THREAD

**Issue by MagicLegend** [gulp-sourcemaps/gulp-sourcemaps#377]
Hi,

I have been fighting with this for three full days now, and I can't get it all to behave like I want it. I am preparing an existing codebase for TS, but creating the correct `tsconfig.json` is being a hell. I have created a separate testing codebase basically from scratch, in which I'll show this issue. It goes further than this, but the main issue now lies in the sourcemaps pointing to incorrect paths. I'm at a total loss :(

**The issue I'm having is that manually invoking the transpiler results in different sourcemaps than using the plugin.**

Files:

*tsconfig.base.json*
```json
{
    "compileOnSave": false,
    "compilerOptions": {
      "target": "es6",
      "module": "commonjs",
      "moduleResolution": "node",
      "experimentalDecorators": true,
      "emitDecoratorMetadata": true,
      "sourceMap": true,
      "pretty": true,
      "strictNullChecks": true,
      "skipLibCheck": true,
      "strictPropertyInitialization": true,
      "outDir": ".nowhere", // Bogus value; this config should not be used to transpile.
      // "rootDir": "src",
      "allowJs": true
    }
}
```
*tsconfig.json*
```json
{
  // https://stackoverflow.com/a/37555279/7193940
  "extends": "./tsconfig.base.json",
  "compilerOptions": {
    "allowJs": true,
    "outDir": "bin"
  },
  "include": [
    "src",
    "test",
    "gulpfile.js"
  ],
  "exclude": [
    "node_modules",
    "bin"
  ]
}
```

*tsconfig.dev.json*
```json
{
  "extends": "./tsconfig.base.json",
  "compilerOptions": {
    "allowJs": true,
    "isolatedModules": true,
    "outDir": "bin",
    "sourceMap": true,
    // "baseUrl": "./"
  },
  "include": [
    "src"
  ],
  "exclude": [
    "node_modules",
    "bin",
    "gulpfile.js"
  ]
}
```

*Gulpfile.js*
```js
const gulp = require('gulp');
const tsc = require('gulp-typescript');
const sourcemaps = require('gulp-sourcemaps');
const del = require('del');
const eslint = require('gulp-eslint');
const logger = require('gulplog');
const notify = require('gulp-notify');
const nodemon = require('gulp-nodemon');
const mocha = require('gulp-mocha');

const tsProject = tsc.createProject('tsconfig.dev.json');

const config = {
    source: './',
    get sourceApp() { return `${this.source}src/`; },
    get tsOutputPath() { return `${this.source}bin/`; },
    get tsTestOutputPath() { return `${this.source}bin/`; },
    get allJavaScript() { return `${this.source}bin/**/*.js`; },
    get allTypeScript() { return `${this.sourceApp}**/*.ts`; },
};

/**
 * Transpile all TypeScript in the project and store the output + sourcemaps in the config.tsOutputPath folder.
 */
gulp.task('transpile-ts', () => tsProject.src()
    .pipe(sourcemaps.init())
    .pipe(tsProject())
    .pipe(sourcemaps.write('.', { sourceRoot: '.', includeContent: false }))
    .pipe(gulp.dest(config.tsOutputPath)));

gulp.task('mocha', () => {
    return gulp.src(['test/**/*.test.ts']).pipe(mocha({ require: 'ts-node/register', reporter: 'nyan' }));
});

/**
 * Utilize nodemon to run the node server.
 */
gulp.task('start-node', () => nodemon({
    // the script to run the app
    script: `${config.tsOutputPath}index.js`,
    // this listens to changes in any of these files/routes and restarts the application
    watch: [config.sourceApp],
    tasks: ['transpile'],
    ext: 'ts,js',
    env: {
        NODE_ENV: 'development',
    },
})
    // .on('start', ['transpile'])
    // .on('change', ['watch-no-lint'])
    .on('restart', () => {
        if (config.DEBUG) {
            gulp.src()
                .pipe(notify('Restarting node server...'));
        }
    }).on('crash', () => {
        gulp.src()
            .pipe(notify('Nodemon has crashed!'));
    }));

gulp.task('transpile', gulp.series('clean-ts', 'transpile-ts'));
```

**When using the `gulp transpile` command:**

*Source location: src/index.ts*
*Transpiled location: bin/index.js*
*Mapping file location: bin/index.js.map*
```json
{"version":3,"sources":["index.ts"],"names":[],"mappings":";;AAAA,wBAAsB","file":"index.js","sourceRoot":"."}
```

**When using `tsc --p tsconfig.dev.json` command:**

*Source location: src/index.ts*
*Transpiled location: bin/index.js*
*Mapping file location: bin/index.js.map*
```json
{"version":3,"file":"index.js","sourceRoot":"","sources":["../src/index.ts"],"names":[],"mappings":";;AAAA,wBAAsB"}
```

**As you can see, the `sources` of the `gulp-transpile` command are plain wrong, but the ones generated by `tsc` aren't.** I'm at a total loss of what I'm doing wrong, and any pointers would be greatly appreciated. The complete project that I'm using for testing can be found here: https://github.com/MagicLegend/ts-testing-project

In the actual project I am also having issues with that all the source files are located in `server/src`, which confuses the source map generation even more (it adds a `server/` to the `sources` path...)

Any help is much appreciated!

**Comment by phated** [comment]
tsProject is known to be broken. Sorry.

**Comment by MagicLegend** [comment]
Hi @phated ! Thanks for the extreme quick reply. Are you aware of any alternative how to make this work then?

**Comment by phated** [comment]
Unfortunately not. I don't use typescript, so I don't use that plugin. Sorry 😞

**Comment by MagicLegend** [comment]
Hmm, that sucks. Maybe at least make a note on the wiki that `gulp-typescript` support is flaky to say the least then.

I finally figured out how to do this; and the solution turned out to be to drop gulp completely. The solution is present in https://github.com/MagicLegend/ts-testing-project, and is as follows:

Add npm script:
```json
"start-watch": "./node_modules/.bin/tsc-watch --onSuccess \"node --inspect bin/index.js\" --onFailure \"notify-send -a Node 'Transpilation error!' 'Check the task terminal for results...'\" -p tsconfig.dev.json"
```

Done. I got the solution from [this SO answer](https://stackoverflow.com/a/47305304/7193940) ([archived](https://web.archive.org/web/20200319153243/https://stackoverflow.com/questions/36619472/run-a-npm-script-from-gulp-task)). The debugger can just be attached; and everything magically works.

Again thank you for your quick response! I hope my solution will help somebody else facing the same problems.
